### DISCUSSION TITLE
Fehler bei Modbus Schalter: Cannot connect to Modbus

### DISCUSSION THREAD

**Discussion by HansESA** [camueller/SmartApplianceEnabler#517] [Installation und Betrieb]
Hallo,
bei meiner Lambda Wärmepumpe EU13L kann ich dank deiner Hilfe nun die Leistung über Modbus TCP auslesen.
Der Lambda Wärmepumpe kann man über Modbus auch den PV-Überschuss mitteilen und die Wärmepumpe entscheidet dann selbst, ob und wieviel Leistung sie dann einsetzt. Dieser Wert muss aller < 5min gesetzt werden, sonst fällt er wieder auf 0 zurück.

Dafür habe ich zwei Geräte im SAE angelegt.
1. Gerät nur Zählen, als Master-Zählen
2. Gerät Zählen als Slave und Stufenschalter (Oder geht es auch stufenlos?) um den PV-Überschuss an die Wärmepumpe zu senden

![grafik](https://user-images.githubusercontent.com/128302792/233854921-bf516ed1-9be4-40e2-b10f-66b97b010b51.png)

![grafik](https://user-images.githubusercontent.com/128302792/233854552-507a0820-b0df-416d-bf57-7d19decf5aec.png)

![grafik](https://user-images.githubusercontent.com/128302792/233855130-cb3863bd-1df2-4450-a688-7a79fc0d1586.png)

Im Log-File kommt nach einem switching on, aber immer cannot connect to modbus. Aber es gibt auch Einträge zum Gerät 2 mit erfolgreichen connect to modbus.
Mit Simply Modbus kann ich problemlos einen PV-Überschuss ins Register schreiben. Im Beispiel mit FC 16 in Register 102 der Wert 2100 (Watt).
Gibt es evt. Schwierigkeiten im SAE da ich bei FC nur 6/16 und nicht allein FC 16 auswählen kann?
![grafik](https://user-images.githubusercontent.com/128302792/233854652-de292513-c73b-4ead-b47f-1a129c5e60aa.png)

Das Log-File würde ich dir gern mal zuschicken.
Ich bin gespannt auf deine / eure Ideen zum Lösen des Problems.

Vorab Danke für die Hilfe.

**Comment by camueller** [comment]
Spontan fällt mir auf, dass Du in Simply Modbus als Slave-Adresse 1 gesetzt hast, aber im SAE 2.
Poste doch mal die Modbus-Konfiguration im SAE ...

**Comment by HansESA** [comment]
Die Wärmepumpe reagiert auf Slave 1 und Slave 2. Wahrscheinlich sogar bis 16, laut Doku können nämlich bis zu 16 Modbus Masters gleichzeitig mit ihr kommunizieren. Anbei Beispiel mit Simply Modbus und Slave 2:
![grafik](https://user-images.githubusercontent.com/128302792/233865371-ade38d20-e0e6-485f-a477-c78c1550f6c0.png)

Modbus Konfiguration im SAE:
![grafik](https://user-images.githubusercontent.com/128302792/233865433-f41ec825-febd-4544-8514-ec78a9cda58b.png)
Zum Auslesen der Leistung funktioniert diese auch sehr gut, sowohl als Slave 1 als auch als Slave 2.

**Comment by camueller** [comment]
Wenn Du mit Simply Modbus per TCP mit der WP kommunizieren kannst, sollte das mit analogen Einstellungen auch im SAE funktionieren. Da kann aus der Ferne aber vermutlich niemand helfen.

> Der Lambda Wärmepumpe kann man über Modbus auch den PV-Überschuss mitteilen und die Wärmepumpe entscheidet dann 
> selbst, ob und wieviel Leistung sie dann einsetzt. Dieser Wert muss aller < 5min gesetzt werden, sonst fällt er wieder auf 0 zurück.

Das wird im SAE nicht funktionieren, weil dieser den Schalter nur schaltet, wenn der Schaltbefehl vom SHM kommt. Ich habe so eine periodische Wiederholung schon einmal für eine Wallbox eingebaut, aber für anderen Schalter kann der SAE das aktuell nicht. Könnte man vielleicht mal einbauen ...

> Dafür habe ich zwei Geräte im SAE angelegt.

Soweit so gut.

> Gerät nur Zählen, als Master-Zählen
> Gerät Zählen als Slave und Stufenschalter (Oder geht es auch stufenlos?) um den PV-Überschuss an die Wärmepumpe zu senden

Ein Stufenschalter ist nur für Geräte sinnvoll, deren Leistung steurbar ist. Das ist bei Dir nicht der Fall: Dein Gerät kann nur ein-/ausgeschaltet werden (in Abhängigkeit vom vorhandenen Überschuss). Aber wegen der nicht unterstützten Periodizität des Schalters ist diese Antwort eher theoretischer Natur.

**Comment by HansESA** [comment]
Eventuell habe ich auch nur was bei den Einstellungen im SAE übersehen. Die IP beim MQTT Broker muss ich nicht anpassen, oder?
![grafik](https://user-images.githubusercontent.com/128302792/234357733-f69648e1-5ab6-4d63-98ad-0677ffbea35d.png)

Bei der Lambda Wärmepumpe kann ich indirekt die Leistung steuern. Der Modbus Stufenschalter enthält je Stufe eine Leistung, die als Einschaltbefehl gesendet wird. Ausschalten erfolgt dann über 0 W. Der SHM soll dann entsprechend des geplanten PV Überschusses die jeweilige Stufe schalten. Die Wärmepumpe nimmt dies als Empfehlung. 
Ähnlich funktioniert es ja mit dem SG Ready Eingang. Im SHM sind 1500 W zum Einplanen hinterlegt, aber tatsächlich kann die Wärmepumpe dann auch weniger verbrauchen.

**Comment by camueller** [comment]
> Eventuell habe ich auch nur was bei den Einstellungen im SAE übersehen. Die IP beim MQTT Broker muss ich nicht anpassen, oder?

Wenn Du Modbus laut SAE-Dokumentation verwendest, dann sollte es so passen.

**Comment by HansESA** [comment]
Hallo,
gestern gab es genügend Sonne. Im Log gibt es also einen Schaltbefehl, allerdings wird die Modbus Verbindung zum Gerät 02 (Schalter) nicht aufgebaut. Was könnte hierfür die Ursache sein? 
Fehler beim Aufbau der Verbindung oder Fehler beim Senden des Schaltbefehles?

![grafik](https://user-images.githubusercontent.com/128302792/235100859-bbc3b1af-5052-4152-a0a1-e925f0489cbd.png)

Die Modbus Verbindung wird von der Wärmepumpe nach einer Minute Inaktivität getrennt und muss daher neu aufgebaut werden.
Da diese Modbus Verbindung für den Zähler in Gerät 01 (nur messen) genutzt wird, ist sie eigentlich immer aktiv.
Als Register Typ kann ich im SAE nur FC=6 / 16 auswählen. Zum Schreiben benötige ich aber unbedingt FC=16, mit FC=6 gibt es einen Fehler. 
Was sendet der SAE tatsächlich, FC=6 oder 16? 
Die zu sendenden Parameter-Werte müssen im SAE als Dezimal angeben werden, nicht als HEX?

Vielen Dank für die Unterstützung und das großartige Projekt.

**Comment by camueller** [comment]
Der SAE verwendet eine Standard-Bibliothekt für Modbus und baut die Verbindung jeweils neu auf bei jdem Befehl oder Abfrage.
Letztlich musst Du im SAE nur dasselbe konigurieren, wie im Simply Modbus. Aber solange der SAE nicht das periodische Widerholen des Schalterzustandes untersützt, brauchen wir keine Zeit mehr für irgendwelche Versuche zu investieren.

**Comment by HansESA** [comment]
Hallo, 
da ich noch immer vom SAE und seine Funktion überzeugt bin, habe ich mir eine Lösung für das periodische Wiederholen des Schalterzustandes ausgedacht. 
Auf dem Raspberry läuft parallel der ioBroker. Der SAE sendet jetzt den Schaltzustand per Modbus nicht mehr an die Wärmepumpe, sondern an den ioBroker. Ein Skript im ioBroker schreibt dann zyklisch diesen Wert über Modbus in die Wärmepumpe. 
Doch leider ist das Grundproblem geblieben. Sobald der SAE einen Schaltbefehl generiert, kommt die Fehlermeldung "Cannot connect to modbus..." Mit Simply Modbus kann ich diesen Wert jedoch problemlos schreiben.

![grafik](https://github.com/camueller/SmartApplianceEnabler/assets/128302792/88d5fa81-4c27-4f5d-904a-a5a46e7ff9ab)

Welche Daten kann ich zur Verfügung stellen, um das Problem zu lösen? Gibt es erfolgreiche Anwendungen mit einem Modbus-Schalter?

**Comment by camueller** [comment]
Meine Wallbox wird problemlos per Modbuis gesteuert vom SAE. Eine Mobus-gesteuerte Wallbox ist letztlich auch nur ein Schalter mit mehr Befehlen als ein/aus.

**Comment by HansESA** [comment]
Guten Abend, da ich das Schreiben von Modbus Registern irgendwie nicht im SAE hinbekomme, suchte ich eine andere Lösung.
Der SAE schaltet jetzt mit dem Stufenschalter GPIO 2 und GPIO 3. Mit dem iOBroker lese ich den Zustand der GPIOs zyklisch aus und sende es ebenfalls zyklisch als Modbus an meine Lambda Wärmepumpe. Soweit funktioniert es gut.
Allerdings braucht die Wärmepumpe mind. 5 Minuten zum Starten. Im Stufenschalter kann ich ich aber keine Einschaltoption aktivieren.
Welche Zeit wird abgewartet, bis der SAE den Schalter wieder auf off setzt? Im Protokoll finde ich Einträge im 2/3-Minuten-Takt.
Ist das eine Fehlplanung vom SHM, obwohl ich eine Mindestlaufzeit von 30 min hinterlegt habe?

![grafik](https://github.com/camueller/SmartApplianceEnabler/assets/128302792/ad6418c5-0765-41a2-b3df-115c95007713)

Besten Dank für die Unterstützung.

**Reply by camueller** [reply]
Die Einschaltoption ist für Stufenschalter nicht verfügbar.

Der SAE setzt nur die Schaltbefehle vom SHM um, die Du in obirgem Log ja bereits markiert hast. Die Frage ist eher, warum der SHM immer kurz nach dem Einschaltbefehl einen Ausschaltbefehl sendet. Eventuell findest Du das über die SEMP-Logs im SHM heraus: https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/ConnectionAssist_DE.md

**Comment by HansESA** [comment]
Inzwischen sendet der SHM die Befehle korrekt. Ein Neustart hatte geholfen.
